### PR TITLE
Add predicate-based buffering operators: bufferUntil, bufferUntilOther, bufferWhile

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

File: src/main/java/reactor/core/publisher/FluxBufferPredicate.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+	}
+	
+	@Override
+	public void subscribe(Subscriber<? super C> s) {
+		C initialBuffer;
+		
+		try {
+			initialBuffer = bufferSupplier.get();
+		} catch (Throwable e) {
+			Operators.error(s, Operators.onOperatorError(e));
```
Comment (akarnokd) [inline src/main/java/reactor/core/publisher/FluxBufferPredicate.java]: Does onOperatorError handle fatal exceptions?

File: src/main/java/reactor/core/publisher/FluxBufferPredicate.java [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+		@Override
+		public void cancel() {
+			Operators.terminate(S, this);
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.setOnce(S, this, s)) {
+				s.request(Long.MAX_VALUE);
```
Comment (akarnokd) [inline src/main/java/reactor/core/publisher/FluxBufferPredicate.java]: There is no need to go unbounded. Just let the downstream request whatever and in onNext, if you don't emit a complete buffer `request(1)`. You can even do it with ConditionalSubscriber so the conditional-fusion enabled sources don't get bogged down by all those 1 element replenishment calls.

File: src/main/java/reactor/core/publisher/FluxBufferPredicate.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+			}
+		}
+
+		@Override
+		public void onNext(T t) {
+			C b;
+			synchronized (this) {
```
Comment (akarnokd) [inline src/main/java/reactor/core/publisher/FluxBufferPredicate.java]: Any reason to synchronize access to buffer?

File: src/main/java/reactor/core/publisher/FluxBufferPredicate.java (hunk 1 above)
Comment (simonbasle) [inline src/main/java/reactor/core/publisher/FluxBufferPredicate.java]: yes, it calls `Exceptions.throwIfFatal` first thing

File: src/main/java/reactor/core/publisher/FluxBufferPredicate.java (hunk 2 above)
Comment (simonbasle) [inline src/main/java/reactor/core/publisher/FluxBufferPredicate.java]: I had that in mind, yes, to request no more than the necessary number of elements needed to trigger a buffer boundary... I went with the minimum viable product for now but I'll see if I can quickly amend the change.

File: src/main/java/reactor/core/publisher/FluxBufferPredicate.java (hunk 3 above)
Comment (simonbasle) [inline src/main/java/reactor/core/publisher/FluxBufferPredicate.java]: You're right, I don't think it should be necessary, as onNext calls are serialized, and `triggerNewBuffer`/`onNext` wouldn't be called concurrently in this operator. mmh that's a bad case of me having started from the `FluxBufferBoundary` code...

**Comment by simonbasle** [comment]
@akarnokd I have taken your remarks into account. @smaldini suggested that we push the requesting even further by doing `request(n)` to the source instead of simply `request(1)` in some instances, but after trying to implement that I think it might have been over-complexifying things...
Using `ConditionalSubscriber` is probably a good next step though.

I'd like to discuss the `while` vs `until` vs "`until` with "cut-before"`: I think that `bufferWhile` dropping the non-matching elements has its use, as it allows for instance to buffer a stream that contains a separator, without polluting the produced buffers with the separator, in a single step. That's conceptually close to what `takeWhile` describes in its javadoc, so I elected to call this use case `bufferWhile`. The case where the separator should be included, albeit in the next buffer, is exposed as `bufferUntil` with a `cutBefore` additional parameter (to be set to `true`).

**Comment by akarnokd** [comment]
https://github.com/akarnokd/RxJava2Extensions/blob/master/src/main/java/hu/akarnokd/rxjava2/operators/FlowableBufferPredicate.java

**Comment by simonbasle** [comment]
@akarnokd yep, saw that. I still think the "drop the separator" option has its use, eg.:

```java
--- "green" --- "red" --- "#" --- "yellow" --- "blue" -|->
            bufferWhile(c -> !"#".equals(c))
--- ["green", "red"] --- ["yellow", "blue"] -|->
```

**Comment by akarnokd** [comment]
Then extend/rework it with that option.

**Comment by simonbasle** [comment]
it is already in as `bufferWhile`. What you exposed as bufferWhile in RxJava2Extras is exposed as `bufferUntil(predicate, cutBefore)` in this PR (where `bufferUntil(predicate, false)` is merely the same as `bufferUntil(predicate)`).

**Comment by simonbasle** [comment]
With @smaldini we've improved on the PR but weren't able to find a satisfying solution for it to `request(n)` from the source, so in the current state it `request(1)`, but is now a `ConditionalSubscriber`. can you have a look @akarnokd? 🙇

**Comment by simonbasle** [comment]
Note: For the request(n) improvement, I had a last ditch idea of incrementing a `SKIP` counter in request and decrementing it in onNext until it reaches 0, at which point onNext resumes requesting 1... Stephane argued, legitimately, that this represented 2 sync points and was bound to encounter side effects between the 2 modifications of the counter. Might still do a follow-up PR to explore and ensure this is not a viable solution.

**Comment by akarnokd** [comment]
I don't understand what you tried to achieve: is it still request(1)-based or fixed prefetch based? I suggest you do request(1) for now. Benchmark it and work out a prefetch based version separately if necessary.

**Comment by smaldini** [comment]
@akarnokd the prefetch is 1-based with conditional subscriber. In our experiments it didn't seem that valuable to use a prefetch. Are you suggesting an experiment using a fixed prefetch to fill the collection (possibly N) and therefore use something like a Dequeue to release them ?

**Comment by akarnokd** [comment]
The code also uses post-complete emission for some reason I don't fully understand. You'd do that if you'd prefetch and route the overlow into different buffers, queued up and then emitting in response to downstream request, even after the main completed and you have filled-in buffers in the queue.

**Comment by simonbasle** [comment]
there's the case where the last buffer has been partially filled, without encountering a boundary-triggering element before the upstream completes. in this case we want to emit the "partial" buffer from `onCompleted`, and it was my understanding that it would be unsafe to do so without using the drain loops from `DrainUtils` (from our discussions with @smaldini)

**Pull request merged**